### PR TITLE
New merge validator.

### LINKED ISSUE #22: Refactor the Merge Validator
`validate_merged_with()` is the custom validator to determines whether or not one Name can be merged with another. The current implementation is a bit confusing, because validation passes if _any_ exception is thrown. 

We may also want to get the `to_be_merged` variable by looking up the Name object by id, instead of creating the `name_id` manually through string interpolation. This will decouple the validator from the implementation details of the name_id.

This is the current implementation.

``` python
def validate_merged_with(value):
    """
    Custom validator for the merged with CharField.
    We don't want to allow users to do an infinite redirect loop b/c of
    merging a record that is already merged with the current record
    """

    try:
        # attempt to set the variables to their values
        to_be_merged = 'nm%07d' % value
        current = str(Name.objects.get(name_id=to_be_merged).merged_with)
        current_merged = str(Name.objects.get(name_id=current).merged_with)
    except Exception:
        # make the comparison variables different to ensure test failure
        current_merged = '1'
        to_be_merged = '2'

    # test if they are looped
    if current_merged == to_be_merged:
        raise ValidationError(
            u'can\'t merge a record into an infinite redirect loop'
        )
```

### PR DESCRIPTION
**Pull request by damonkelley**
Will resolve #22

### CODE REVIEW

File: name/models.py [hunk 1]
```diff
@@ -190,20 +190,36 @@ def validate_merged_with(value):
...
+    # If we don't find duplicates, then no loop has been created and
+    # the generator will raise it's own StopIteration and we will implicitly
+    # return.
+    merged_list = list()
+    for name in follow_merged_into(merge_target):
+        if name in merged_list:
+            raise ValidationError(u'The specified merge action completes ' +
+                                  'a merge loop. Unable to complete merge.')
+        else:
+            merged_list.append(name)
```
Comment (damonkelley) [inline name/models.py]: This does not need to be in an else condition. It can go directly under the for block.

File: name/models.py [hunk 2]
```diff
@@ -190,20 +190,36 @@ def validate_merged_with(value):
...
+                yield merged_into
+            name = merged_into
+
+    # Iterate through the generator and keep track of the return names.
+    # we will find a loop if the the return name is already in
+    # merged_list. If this happens we will raise a validation error.
+    # If we don't find duplicates, then no loop has been created and
+    # the generator will raise it's own StopIteration and we will implicitly
+    # return.
+    merged_list = list()
```
Comment (damonkelley) [inline name/models.py]: Let's give this a better variable name.

File: name/models.py [hunk 3]
```diff
@@ -190,20 +190,36 @@ def validate_merged_with(value):
...
-    # test if they are looped
-    if current_merged == to_be_merged:
-        raise ValidationError(
-            u'can\'t merge a record into an infinite redirect loop'
-        )
+        merge_target = Name.objects.get(id=id)
+    except Name.DoesNotExist:
+        raise ValidationError('The merge target must exist.')
+
+    def follow_merged_into(name):
```
Comment (damonkelley) [inline name/models.py]: Look at better name for this generator.

File: name/models.py [hunk 4]
```diff
@@ -190,20 +190,35 @@ def validate_merged_with(value):
     """
```
Comment (damonkelley) [inline name/models.py:199]: Update the docstring as well

File: tests/test_validators.py [hunk 5]
```diff
@@ -37,3 +38,39 @@ def test_validate_merged_with_fails(merge_candidates):
 
     with pytest.raises(ValidationError):
         validate_merged_with(primary.id)
+
+
+@pytest.mark.django_db
+def test_validate_merged_fails_when_name_merges_with_itself():
```
Comment (damonkelley) [inline tests/test_validators.py]: This shoud be named `test_validate_merged_with_fails_when_name_merges_with_itself()`

File: name/models.py (hunk 3 above)
Comment (damonkelley) [inline name/models.py]: Scratch that. I am ok with the generator name.

**Comment by damonkelley** [comment]
@ldko, what are your thoughts on this?

File: name/models.py [hunk 6]
```diff
@@ -182,28 +182,52 @@ def __unicode__(self):
...
 
-def validate_merged_with(value):
-    """
-    Custom validator for the merged with CharField.
-    We don't want to allow users to do an infinite redirect loop b/c of
-    merging a record that is already merged with the current record
+def validate_merged_with(id):
+    """Validator for the merged_with ForeignKey field.
+
+    This will look prevent two scenarios from occurring.
```
Comment (ldko) [inline name/models.py]: there are a couple of typos in this docstring

File: name/models.py [hunk 7]
```diff
@@ -182,28 +182,52 @@ def __unicode__(self):
...
+    1. Merging with a nonexistent Name object.
+
+    2. Creating a loop of foreign key relationships.
+        For example:
+            Name 1 -> Name 2 -> Name 3 -> Name 1
+
+        We need to prevent this because navigating to a name that has
+        been merge with another, will redirect you to the Name it has
```
Comment (ldko) [inline name/models.py]: typo in this line and next

**Comment by damonkelley** [comment]
Good catch. Typos should now be eliminated.

File: name/models.py [hunk 8]
```diff
@@ -182,28 +182,52 @@ def __unicode__(self):
...
+        does not have a merged_with relationship.
+        """
+        while name:
+            merged_into = name.merged_with
+            if merged_into:
+                yield merged_into
+            name = merged_into
+
+    # Iterate through the generator and keep track of the return names.
+    # we will find a loop if the return name is already in
```
Comment (ldko) [inline name/models.py:220]: Do you want "we" capitalized?  I know you like your capitalization and punctuation.

Comment (damonkelley) [inline name/models.py:220]: Yes!

I pushed the fix directly to master.

**Pull request merged**